### PR TITLE
Fix issue #48345

### PR DESCRIPTION
**Pull request by kg**
The return value marshaling for bound static methods accidentally passed a root object to the conversion APIs instead of passing the pointer stored inside the root. This would work for many scenarios (intentionally) but not here.

Also renamed _unbox_mono_obj_rooted and _mono_array_to_js_array_rooted to make it clearer what argument type they expect (roots) and separate them from the other _rooted APIs (which do expect raw pointers).

### CODE REVIEW

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs [hunk 1]
```diff
@@ -66,6 +66,12 @@ private static void InvokeString2(string s)
             _stringResource2 = s;
         }
 
+        private static string Issue48345(string s)
+        {
+            _stringResource = $"s: {s} length: {s?.Length}";
```
Comment (kjpou1) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:71]: Nit: This has null check on the `s` variable but do not see a test for `null`.

Comment (kg) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:71]: Doesn't need one, string format on a null will just be ""

Comment (kjpou1) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs:71]: 👍

**Review by kjpou1 — APPROVED** [review]
Looks good

**Review by lewing — COMMENTED** [review]
Fix looks good.  I don't like that the helper and the test share the same name and I'm not sure but I think the we avoid naming tests by issue number but I'm not sure it is blocking.

**Comment by kg** [comment]
> 
> 
> Fix looks good. I don't like that the helper and the test share the same name and I'm not sure but I think the we avoid naming tests by issue number but I'm not sure it is blocking.

I can try and pick a good name for it, the bug is kind of weird so one didn't come to mind.

**Pull request merged**